Add Boosting Mobile Session Replay Performance Post (#167)
getsentry/sentry.engineering: data/blog/boosting-mobile-session-replay-performance-with-view-renderer-v2.md

---

> title: 'Boosting Mobile Session Replay Performance with View Renderer V2'

Author: I changed it to "Boosting Session Replay Performance on iOS with View Renderer V2

Main thread blocking time improved from ~155 milliseconds to ~25 milliseconds per frame, a **~80% reduction** that brings Session Replay overhead well within acceptable performance budgets even on resource-constrained devices.

For teams interested in more detailed information, the complete analysis results are publicly available in [pull request #4940](https://github.com/getsentry/sentry-cocoa/pull/4940) on GitHub.

The optimized view renderer was introduced through a careful rollout process designed to minimize risk while maximizing the performance benefits for users experiencing issues.

Starting with Sentry Cocoa SDK [v8.48.0](https://github.com/getsentry/sentry-cocoa/releases/tag/8.48.0), the new implementation was available as an experimental feature controlled by the `options.sessionReplay.enableExperimentalViewRenderer` flag.

This experimental approach allowed early adopters to test the new renderer in their production environments. Based on positive feedback and telemetry data, the optimized renderer became the default implementation in [v8.50.2](https://github.com/getsentry/sentry-cocoa/releases/tag/8.50.2).

# What's Next

> We've already received new reports of users challenging our new implementation, especially when working with graphics heavy animations.

Reviewer: ```suggestion
While the new renderer improves performance in a vast majority of scenarios, we've already received new reports indicating specific limitations of this new implementation, especially when working with graphics heavy animations.
```

So we went on the journey to find the culprit and found a solution that yielded **4-5x better performance** in our benchmarks 🔎📈

![Session Replay on Mobile](/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/hero.jpg)

To understand what's happening under the hood of Session Replay on Mobile, let's first explore how mobile screen recording works before diving into the technical details.

# Frame Rates In A Nutshell

A screen recording is a video made up of many fast-displayed images called `frames`. Human eyes can process about 60 frames per second (`frame rate`), measured in hertz (`1 Hz` = `1 unit per second`), creating the illusion of a moving image. The frame rates vary by use case, from `24 Hz` for cinema to `144 Hz` for gaming computer displays.

Higher frame rates make videos smoother, but have downsides:

- More storage and network bandwidth required
- Less time to capture per frame available, requiring better hardware

> When reducing the frame rate to a minimum, the recordings look like **stop motion videos**, with a great example for this style visible in this [YouTube video](https://www.youtube.com/watch?v=MEglOulvgSY). The frames are just sequential photos, yet they still feel like a moving photo — essentially a video.

Reviewer 2: ```suggestion
When reducing the frame rate to a minimum, the recordings look like **stop motion videos**, with a great example of this style visible in this [YouTube video](https://www.youtube.com/watch?v=MEglOulvgSY). The frames are just sequential photos, yet they still feel like a moving photo — essentially a video.
```

**This is essentially what we are doing with Session Replay on Mobile!**

> Instead of having a full screen recording, which is resource-heavy, we capture a screenshot every second instead. Every 5 seconds, these frames are combined into a video segment, creating a stop-motion-like recording. These video segments are then uploaded to Sentry and stitched into a full session replay.

Reviewer 2: "Instead of having a full screen recording, which is resource-heavy, we capture a screenshot every second instead."

Now I want to know the `hertz` equivalent measurement for Session Replay :D

Now that we know how frame rates work, let’s dive deeper into the actual problem we had to tackle.

# The Frame Is The Problem

Our investigation began with reproducing the performance issues in a controlled environment. We used an iPhone 8 running iOS 15.7 as our primary test device, since this represented the older hardware where users reported the most severe performance problems.

Using Xcode Instruments on our Sentry Cocoa SDK sample application, we immediately reproduced the reported performance issues and noticed a consistent pattern of main thread hangs **every second**.

<div align="center">
  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/xcode-instruments-app-hangs.png" alt="Xcode Instruments showing an app hang warning every second"/>
  _Xcode Instruments showing an app hang warning every second._
</div>

<div align="center">
  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/coincidence-meme.png" alt="My reaction to the Xcode Instruments report"/>
  _My reaction to the Xcode Instruments report_
</div>

> On iOS each application has exactly one thread responsible for handling the entire UI view hierarchy - the main thread. When view changes occur, the hierarchy is processed by the system's render service, which converts the logical view structure into pixel data for display on the screen.

Reviewer 2: ```suggestion
On each iOS application, there's exactly one thread responsible for handling the entire UI view hierarchy - the main thread. When view changes occur, the hierarchy is processed by the system's render service, which converts the logical view structure into pixel data for display on the screen.
```

Eventually Apple introduced _ProMotion_ displays which adjust their refresh rates up to `120Hz` during interaction and down to `10Hz` when idle — the frame rate is not constant anymore.

> To better understand the exact implications for time-to-render per frame in milliseconds (`ms`), consider that to hit a refresh rate of `120 Hz` the time per frame is narrow with `1000 ms / 120 = ~8.3 ms` available to update and render the view hierarchy. In contrast for `60 Hz` we can double the available time to `1000 ms / 60 = ~16.7 ms` .

Reviewer 2: ```suggestion
To better understand the exact implications for time-to-render per frame in milliseconds (`ms`), consider that to hit a refresh rate of `120 Hz` the time per frame is narrow with `1000 ms / 120 = ~8.3 ms` available to update and render the view hierarchy. In contrast for `60 Hz`, we can double the available time to `1000 ms / 60 = ~16.7 ms` .
```

If the workload is below that, e.g. `4 ms`, all is good and we wait for the next update. But if the workload takes longer than that, e.g. `25ms`, we have a problem — the main thread is blocked and the app hangs.

As animations are not coupled to frames but to time, the app cannot simply continue with rendering the next one afterwards. For example, if you expect a loading indicator to rotate exactly once per second, longer frame render times would slow down the animation.

To correct the timing the system **skips** the frames that should already have been rendered at this point in time (the _dropped frames)_. This correction behavior is commonly known as _frame drops_ and has the undesired side effect of intermittent frame rate changes that can be noticed as visual stutters (i.e. not scrolling smoothly).

<div align="center">
  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/frame-drops.jpg" alt="Dropped Frames"/>
  _When workload takes longer than available time, we need to drop frames_
</div>

> Frame drops were also the unintended side effect for the previous implementation of Session Replay, as the once-per-second screenshot simply took too long.

Reviewer 2: ```suggestion
Frame drops were also the unintended side effect of the previous implementation for Session Replay, as the once-per-second screenshot simply took too long.
```

3. **Mask**: Apply privacy masking by combining the rendered image with the redaction geometry.

As mentioned before, accessing the view hierarchy must be done on the main thread, which means that both steps _Redact_ and _Render_ must also execute on the main thread. Only the final step _Mask_ can be performed on a background queue, since it operates solely on the image data produced by the previous steps.

This allows us to focus on the performance of the _Redact_ and _Render_ steps.

<div align="center">
  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/uikit-meme.png" alt="UIKit Meme"/>
  _Source: [@ios_memes on X](https://x.com/ios_memes/status/1483418141871030276)_
</div>

# Measuring Baseline Performance

After adding some code to calculate the execution time and adding a basic sampling mechanism to reduce variation over time, we quickly gathered our first insights using a (rather old) iPhone 8:

**The step _Render_ takes `~155ms` per frame, causing ~9-10 of 60 frames being dropped every second of using Session Replay** 💀

> | 120 samples | Redact        | Render          | Total           |

Reviewer 2: The totals don't quite add up, but I'm assuming this is because I can only see/add up 4 decimal places?

| ----------- | ------------- | --------------- | --------------- |
| Min         | 3.0583 ms     | 145.3815 ms     | 151.4525 ms     |
| Avg         | 5.8453 ms     | 149.8243 ms     | 155.6732 ms     |
| p50         | 6.0484 ms     | 149.2103 ms     | 154.1397 ms     |
| p75         | 6.1136 ms     | 151.9487 ms     | 158.0255 ms     |

> | p95         | **6.2567 ms** | **155.3496 ms** | **161.3549 ms** |

Reviewer 2: Is it possible to make these measurements a different colour? The bolding is so subtle, I almost missed it :o

Author: Changed it to be bold and in the brand color.

| Max         | 6.5138 ms     | 155.8338 ms     | 161.8351 ms     |

As the duration of _Redact_ with `~6.3ms` is comparatively small, we will optimize it in the future and focus on improving _Render_.

# Optimizing the View Renderer

> Our investigation revealed that the rendering phase consumed approximately 155 milliseconds per screenshot, representing the primary performance bottleneck. The original implementation relied on Apple's high-level UIGraphicsImageRenderer API, which provides convenient abstractions but introduces significant overhead for our use case.

Reviewer 2: ```suggestion
Our investigation revealed that the rendering phase consumed approximately 155 milliseconds per screenshot, representing the primary performance bottleneck. The original implementation relied on Apple's high-level `UIGraphicsImageRenderer` API, which provides convenient abstractions but introduces significant overhead for our use case.
```

Here's the baseline implementation that was causing the performance issues:

```swift
func render(view: UIView) -> UIImage {
    // Setup: Create graphics context and configure rendering environment
    let screenshot = UIGraphicsImageRenderer(size: view.bounds.size).image { context in
        // Draw: Render view hierarchy into the graphics context
        view.drawHierarchy(in: view.bounds, afterScreenUpdates: false)
    }
    return screenshot
}
```

This implementation consists of two primary function blocks that each present optimization opportunities.

> - **Setup**: Creates a graphical bitmap context that serves as the "canvas" for rendering operations, then converts the resulting bitmap data into a UIImage after drawing completes.

Reviewer 2: ```suggestion
- **Setup**: Creates a graphical bitmap context that serves as the "canvas" for rendering operations, then converts the resulting bitmap data into a `UIImage` after drawing completes.
```

<div align="center">
  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/render-comparison.png" alt="Render Failure"/>
  _Tab bar icons missing when using layer.render(in:) approach_
</div>

The impact of the missing UI is not entirely clear and we could not pinpoint the exact cause for this (with unverified assumptions it is a side effect of [SF Symbols](https://developer.apple.com/sf-symbols/) and the way fonts are rendered). We decided against adopting this behavior as the new default render method in the foreseeable future and stick with `drawHierarchy` instead.

As some of you might still prefer the even faster render times over completely rendered frames, you can still opt-in by configuring the option `options.sessionReplay.enableFastViewRenderering = true`.

# The Numbers Speak For Themselves

The optimization achieved substantial performance improvements, with the most dramatic gains on older hardware. Frame drops decreased from 9-10 frames per second to approximately 2 frames per second on iPhone 8 devices, representing a massive win!

> Main thread blocking time improved from 155 milliseconds to 25 milliseconds per screenshot, an **84% reduction** that brings Session Replay overhead well within acceptable performance budgets even on resource-constrained devices.

Reviewer 2: We say "80%" above if I understand correctly, so this seems a _little_ inconsistent ("We have our first win and reduced the average time by ~125ms or ~80%! 🎉")
